cli/npm/resolution/graph.rs

```diff
@@ -184,31 +192,34 @@ impl NodeIdRef {
 #[derive(Clone)]
 enum GraphPathNodeOrRoot {
   Node(Arc<GraphPath>),
-  Root(NpmPackageNv),
+  Root(Arc<NpmPackageNv>),
 }
 
 /// Path through the graph that represents a traversal through the graph doing
 /// the dependency resolution. The graph tries to share duplicate package
 /// information and we try to avoid traversing parts of the graph that we know
 /// are resolved.
-#[derive(Clone)]
 struct GraphPath {
   previous_node: Option<GraphPathNodeOrRoot>,
   node_id_ref: NodeIdRef,
-  // todo(dsherret): I think we might be able to get rid of specifier and
-  // node version here, but I added them for extra protection for the time being.
   specifier: String,
-  nv: NpmPackageNv,
+  // we could consider not storing this here and instead reference the resolved
+  // nodes, but we should performance profile this code first
+  nv: Arc<NpmPackageNv>,
+  /// Descendants in the path that circularly link to an ancestor in a child.These
+  /// descendants should be kept up to date and always point to this node.
+  linked_circular_descendants: Mutex<Vec<Arc<GraphPath>>>,
```

Comment:
This is the new concept here. We keep track of descendants that have this ancestor as a child.